Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp

```diff
@@ -89,257 +89,358 @@ void ADefiWalletCoreActor::Destroyed() {
+            output_message =
+                FString::Printf(TEXT("CronosPlayUnreal GetNFTToken Error: %s"),
+                                UTF8_TO_TCHAR(e.what()));
         }
-        TxDirector director;
-        CosmosSendAmountTxBuilder builder;
-        director.setBuilder(&builder);
-        CosmosSDKTxInfoRaw tx_info = director.makeTx();
-
-        assert(NULL != _coreWallet);
-        std::string mychainid = TCHAR_TO_UTF8(*myChainID);
-        std::string myfrom = TCHAR_TO_UTF8(*fromaddress);
-        std::string myto = TCHAR_TO_UTF8(*toaddress);
-        uint64 myamount = (uint64)amount;
-        std::string myamountdenom = TCHAR_TO_UTF8(*amountdenom);
-        std::string myservercosmos =
-            TCHAR_TO_UTF8(*myCosmosRpc); /* 1317 port */
-        std::string myservertendermint =
-            TCHAR_TO_UTF8(*myTendermintRpc); /* 26657 port */
-        UE_LOG(LogTemp, Log,
-               TEXT("CronosPlayUnreal SendAmount from %s to %s amount %lld %s"),
-               UTF8_TO_TCHAR(myfrom.c_str()), UTF8_TO_TCHAR(myto.c_str()),
-               myamount, UTF8_TO_TCHAR(myamountdenom.c_str()));
 
-        rust::cxxbridge1::String result =
-            _coreWallet->get_default_address(CoinType::CryptoOrgMainnet);
-
-        rust::cxxbridge1::String detailjson =
-            query_account_details(myservercosmos, myfrom);
-
-        org::defi_wallet_core::CosmosAccountInfoRaw detailinfo =
-            query_account_details_info(myservercosmos, myfrom);
-        tx_info.account_number = detailinfo.account_number;
-        tx_info.sequence_number = detailinfo.sequence_number;
-        tx_info.chain_id = mychainid;
-        char hdpath[100];
-        // m / purpose' / coin_type' / account' / change / address_index
-        snprintf(hdpath, sizeof(hdpath), "m/44'/%d'/0'/0/%d", tx_info.coin_type,
-                 walletIndex);
-        rust::cxxbridge1::Box<org::defi_wallet_core::PrivateKey> privatekey =
-            _coreWallet->get_key(hdpath);
+        AsyncTask(ENamedThreads::GameThread,
+                  [Out, output, success, output_message]() {
+                      Out.ExecuteIfBound(output, success, output_message);
+                  });
+    });
+}
 
-        rust::cxxbridge1::Vec<uint8_t> signedtx =
-            get_single_bank_send_signed_tx(tx_info, *privatekey, myto, myamount,
-                                           myamountdenom);
+rust::cxxbridge1::Vec<uint8_t> ADefiWalletCoreActor::MakeTxForSendAmount(
+    int32 walletIndex, FString fromaddress, FString toaddress, int64 amount,
+    FString amountdenom) {
+    TxDirector director;
+    CosmosSendAmountTxBuilder builder;
+    director.setBuilder(&builder);
+    CosmosSDKTxInfoRaw tx_info = director.makeTx();
+
+    assert(NULL != _coreWallet);
+    std::string mychainid = TCHAR_TO_UTF8(*myChainID);
+    std::string myfrom = TCHAR_TO_UTF8(*fromaddress);
+    std::string myto = TCHAR_TO_UTF8(*toaddress);
+    uint64 myamount = (uint64)amount;
```

Comment:
is this cast safe? should there be an assertion that amount is non-negative?

Reply:
yes, it can raise overflow issue, i'll add exception handling